diff --git a/src/fandango/language/tree.py b/src/fandango/language/tree.py
index bc562b7a..8972dd6b 100644
--- a/src/fandango/language/tree.py
+++ b/src/fandango/language/tree.py
@@ -324,12 +324,13 @@ class DerivationTree:
         self.invalidate_hash()
 
     def find_subtrees(
-        self, symbol: NonTerminal | str
+        self, symbol: NonTerminal | str, stop_at: Optional[NonTerminal] = None, bfs_mode: bool = True
     ) -> Generator["DerivationTree", None, None]:
         """
         Recursive, breadth-first search to find all trees with the given non-terminal symbol under this tree, including sources of generators.
 
         :param symbol: The non-terminal symbol to find.
+        :param stop_at: An optional actual non-terminal at which to stop the search
         :return: A generator of all trees with the given non-terminal symbol under this tree.
         """
         if isinstance(symbol, str):
@@ -339,9 +340,18 @@ class DerivationTree:
         queue = deque([self])
         while queue:
             current = queue.popleft()
+            if stop_at is not None and current is stop_at:
+                # End the whole thing
+                break
             if current.symbol.is_non_terminal and current.symbol == symbol:
                 yield current
-            queue.extend([*current._children, *current._sources])
+            # for BFS, we add children and sources to the end of the queue
+            if bfs_mode:
+                queue.extend([*current._children, *current._sources])
+            # for DFS, we add children and sources to the beginning of the queue
+            else:
+                queue.extendleft(reversed(current._sources))
+                queue.extendleft(reversed(current._children))
 
     def find_all_trees(self, symbol: NonTerminal | str) -> list["DerivationTree"]:
         warnings.warn(
diff --git a/src/fandango/language/tree_value.py b/src/fandango/language/tree_value.py
index d6b27041..afea8366 100644
--- a/src/fandango/language/tree_value.py
+++ b/src/fandango/language/tree_value.py
@@ -36,7 +36,7 @@ def _str_to_bytes(value: str, encoding: str) -> bytes:
     :return: A bytes object.
     """
     try:
-        return value.encode(encoding=encoding)
+        return value.encode(encoding=encoding, errors="surrogatepass")
     except UnicodeEncodeError as e:
         raise FandangoConversionError(
             f"string to bytes conversion failed, string: {value}, encoding: {encoding}, error: {e}"
@@ -53,7 +53,7 @@ def _bytes_to_str(value: bytes, encoding: str) -> str:
     :return: A string.
     """
     try:
-        return value.decode(encoding=encoding)
+        return value.decode(encoding=encoding, errors="surrogatepass")
     except UnicodeDecodeError as e:
         raise FandangoConversionError(
             f"bytes to string conversion failed, bytes: {value!r}, encoding: {encoding}, error: {e}"
